File: python/zensical/compat/mkdocstrings.py
# Copyright (c) 2025-2026 Zensical and contributors

# SPDX-License-Identifier: MIT
# All contributions are certified under the DCO

# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to
# deal in the Software without restriction, including without limitation the
# rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
# sell copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:

# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.

# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
# IN THE SOFTWARE.

from __future__ import annotations

from dataclasses import dataclass
from io import BytesIO
from pathlib import Path
from typing import TYPE_CHECKING, Any, cast

from zensical.extensions.autorefs import get_autorefs_store

if TYPE_CHECKING:
    from mkdocstrings import (
        Handlers,
        MkdocstringsExtension,
    )


# ----------------------------------------------------------------------------
# Globals
# ----------------------------------------------------------------------------


HANDLERS: Handlers | None = None
_ENABLE_INVENTORY: bool | None = None


# ----------------------------------------------------------------------------
# Classes
# ----------------------------------------------------------------------------


class ToolConfig:
    """Mock mkdocstrings tooling configuration."""

    def __init__(self, config_file_path: str | None = None) -> None:
        self.config_file_path = config_file_path


@dataclass(frozen=True, order=True)
class _SortableBacklinkCrumb:
    """Backlink crumb with title-aware equality and ordering."""

    title: str
    url: str


# ----------------------------------------------------------------------------
# Functions
# ----------------------------------------------------------------------------


def _python_handler_has_backlinks(
    handlers: dict[str, Any] | None,
) -> bool:
    """Return whether the Python handler enables backlinks."""
    if not isinstance(handlers, dict):
        return False
    handler = handlers.get("python")
    if not isinstance(handler, dict):
        return False
    options = handler.get("options")
    return isinstance(options, dict) and bool(options.get("backlinks", False))


def _get_handlers(
    handlers: dict[str, Any] | None = None,
    *,
    custom_templates: str | None = None,
    enable_inventory: bool | None = None,
    default_handler: str = "python",
    locale: str = "en",
    config: dict[str, Any],
) -> Handlers:
    """Create or return the handlers shared by the current build."""
    from mkdocstrings import (  # noqa: PLC0415
        Handlers,
    )

    global HANDLERS, _ENABLE_INVENTORY  # noqa: PLW0603
    _ENABLE_INVENTORY = enable_inventory
    if HANDLERS is None:
        root_dir = Path(config["root_dir"])
        config_file = root_dir / "zensical.toml"
        tool_config = ToolConfig(config_file_path=str(config_file))
        HANDLERS = Handlers(
            theme="material",
            default=default_handler,
            inventory_project=config["site_name"],
            inventory_version="0.0.0",
            handlers_config=handlers if handlers is not None else {},
            custom_templates=custom_templates,
            mdx=config["markdown_extensions"],
            mdx_config=config["mdx_configs"],
            locale=locale,
            tool_config=tool_config,
        )

        HANDLERS._download_inventories()
        url_map = get_autorefs_store()._abs_url_map
        for identifier, url in HANDLERS._yield_inventory_items():
            url_map[identifier] = url

    return HANDLERS


def _ensure_handlers() -> Handlers | None:
    """Initialize handlers when every Markdown page came from cache."""
    if HANDLERS is not None:
        return HANDLERS

    from zensical.config import get_config  # noqa: PLC0415

    config = get_config()
    options = config["mdx_configs"].get("zensical.extensions.mkdocstrings")
    if options is None or not options.get("enabled", True):
        return None
    options = dict(options)
    options.pop("enabled", None)
    return _get_handlers(**options, config=config)


def _get_backlink_handler(handler_name: str) -> tuple[Handlers, Any] | None:
    """Return a handler prepared to render backlinks on cached builds."""
    handlers = _ensure_handlers()
    if handlers is None:
        return None

    handler = handlers.get_handler(handler_name)
    if getattr(handler, "_md", None) is None:
        from markdown import Markdown  # noqa: PLC0415

        handler._update_env(Markdown(), config=handlers._tool_config)
    return handlers, handler


def get_mkdocstrings_extension(
    handlers: dict[str, Any] | None = None,
    *,
    custom_templates: str | None = None,
    enable_inventory: bool | None = None,
    default_handler: str = "python",
    locale: str = "en",
    config: dict[str, Any],
) -> MkdocstringsExtension:
    """Create the mkdocstrings Markdown extension."""
    from mkdocstrings import (  # noqa: PLC0415
        MkdocstringsExtension,
    )

    autorefs = get_autorefs_store()
    autorefs.record_backlinks = _python_handler_has_backlinks(handlers)
    handlers_instance = _get_handlers(
        handlers,
        custom_templates=custom_templates,
        enable_inventory=enable_inventory,
        default_handler=default_handler,
        locale=locale,
        config=config,
    )
    # Upstream annotates this as its full plugin; our store supplies the
    # compatible anchor-registration interface used by the extension.
    return MkdocstringsExtension(
        handlers=handlers_instance, autorefs=cast("Any", autorefs)
    )


def get_inventory(cached: bytes | None) -> bytes:
    """Get inventory bytes, merging cached entries with fresh handlers data."""
    if HANDLERS is None:
        return cached or b""

    try:
        from mkdocstrings import (  # noqa: PLC0415
            Inventory,
        )
    except ImportError:
        return cached or b""

    if not cached:
        return HANDLERS.inventory.format_sphinx()

    base = Inventory.parse_sphinx(BytesIO(cached))
    for name, item in HANDLERS.inventory.items():
        base[name] = item

    # Bug in mkdocstrings's `parse_sphinx` method
    # not parsing project and version (fixed in latest).
    base.project = HANDLERS.inventory.project
    base.version = HANDLERS.inventory.version

    return base.format_sphinx()


def get_backlink_aliases(handler_name: str, identifier: str) -> list[str]:
    """Return aliases used by a handler for a backlink placeholder."""
    result = _get_backlink_handler(handler_name)
    if result is None:
        return []
    _, handler = result
    return list(handler.get_aliases(identifier))


def render_backlinks(
    handler_name: str,
    backlinks: list[tuple[str, list[list[tuple[str, str]]]]],
) -> str:
    """Turn serializable backlink crumbs into handler-rendered HTML."""
    if not backlinks:
        return ""
    result = _get_backlink_handler(handler_name)
    if result is None:
        return ""
    handlers, handler = result

    from inspect import signature  # noqa: PLC0415

    from mkdocs_autorefs import (  # noqa: PLC0415
        Backlink,
    )

    # Rust already returns these paths sorted and deduplicated. Preserve that
    # order instead of introducing Python's process-random set order again.
    # Our crumbs supply the title and URL fields expected by the handler.
    data = {
        backlink_type: tuple(
            Backlink(
                cast(
                    "Any",
                    tuple(
                        _SortableBacklinkCrumb(title=title, url=url)
                        for title, url in crumbs
                    ),
                )
            )
            for crumbs in backlink_list
        )
        for backlink_type, backlink_list in backlinks
    }
    kwargs = {}
    if "locale" in signature(handler.render_backlinks).parameters:
        kwargs["locale"] = handlers._locale
    return handler.render_backlinks(data, **kwargs)


def get_inventory_policy(cached_auto_enabled: bool) -> tuple[bool, bool]:
    """Decide whether to write `objects.inv`.

    Return two booleans: whether to write the file, and whether any handler
    enables it by default.

    We remember the second value for cached pages, whose handlers may not run
    again. We reuse it only if the project configuration has not changed.
    """
    from zensical.config import get_config  # noqa: PLC0415

    config = get_config()
    plugin = config.get("plugins", {}).get("mkdocstrings", {}).get("config", {})
    options = config.get("mdx_configs", {}).get(
        "zensical.extensions.mkdocstrings", plugin
    )
    auto_enabled = cached_auto_enabled or (
        HANDLERS is not None
        and any(handler.enable_inventory for handler in HANDLERS.seen_handlers)
    )
    setting = (
        _ENABLE_INVENTORY
        if HANDLERS is not None
        else options.get("enable_inventory")
    )
    enabled = options.get("enabled", True) and (
        auto_enabled if setting is None else setting
    )
    return enabled, auto_enabled


def reset() -> None:
    """Reset global state in-between rebuilds."""
    global HANDLERS, _ENABLE_INVENTORY  # noqa: PLW0603
    HANDLERS = None
    _ENABLE_INVENTORY = None

File: crates/zensical/src/compat/mkdocs/plugin/mkdocstrings.rs
// Copyright (c) 2025-2026 Zensical and contributors

// SPDX-License-Identifier: MIT
// All contributions are certified under the DCO

// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to
// deal in the Software without restriction, including without limitation the
// rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
// sell copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:

// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.

// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
// IN THE SOFTWARE.

// ----------------------------------------------------------------------------

//! Mkdocstrings compatibility plugin.

use pyo3::types::PyAnyMethods;
use pyo3::Python;
use serde::de::DeserializeOwned;
use serde::{Deserialize, Serialize};
use std::fs;
use std::hash::{DefaultHasher, Hash, Hasher};
use std::path::PathBuf;

use zrx::id::Id;
use zrx::stream::Signal;

use crate::compat::mkdocs::html::Editor;
use crate::compat::mkdocs::plugin::autorefs::Registry;
use crate::config::Config;
use crate::path::{OutputRoot, SitePath};
use crate::structure::nav::Navigation;

mod backlinks;

pub use backlinks::Parser;

// ----------------------------------------------------------------------------
// Constants
// ----------------------------------------------------------------------------

/// Python Markdown extension that enables mkdocstrings compatibility.
const EXTENSION_NAME: &str = "zensical.extensions.mkdocstrings";

/// Version of cached backlink page data.
const BACKLINK_CACHE_VERSION: u8 = 3;

// ----------------------------------------------------------------------------
// Structs
// ----------------------------------------------------------------------------

/// On-disk cache for backlink data grouped by rendered page.
#[derive(Clone, Debug)]
struct BacklinkCache {
    /// Directory containing independently addressable page data.
    directory: PathBuf,
    /// Hash of all configuration that can affect handler rendering.
    config_hash: u64,
}

// ----------------------------------------------------------------------------

/// One page artifact paired with the hash of its inputs.
#[derive(Debug, Deserialize, Serialize)]
struct Cached<T> {
    /// Cached artifact.
    data: T,
    /// Hash of configuration and computation arguments.
    hash: u64,
}

// ----------------------------------------------------------------------------

/// Mkdocstrings compatibility pipeline.
#[derive(Clone, Debug)]
pub struct Mkdocstrings {
    /// Cache directory shared with the Python compatibility layer.
    cache: PathBuf,
    /// Backlink cache, created only when collection is enabled.
    backlink_cache: Option<BacklinkCache>,
    /// Site output directory.
    output: OutputRoot,
}

// ----------------------------------------------------------------------------

/// Inputs required to generate the object inventory.
pub struct Dependencies<'a> {
    /// Revision-complete site navigation.
    pub navigation: &'a Signal<Id, Navigation>,
}

// ----------------------------------------------------------------------------
// Implementations
// ----------------------------------------------------------------------------

impl BacklinkCache {
    /// Returns cached page data, computing and storing it on a miss.
    fn get_or_compute<T, I, A, F>(
        &self, namespace: &str, id: &I, args: &A, compute: F,
    ) -> anyhow::Result<T>
    where
        T: DeserializeOwned + Serialize,
        I: Hash + ?Sized,
        A: Hash + ?Sized,
        F: FnOnce() -> anyhow::Result<T>,
    {
        let path = self.path(namespace, id);
        let hash = self.hash(id, args);
        if let Ok(data) = fs::read(&path)
            && let Ok(cached) = serde_json::from_slice::<Cached<T>>(&data)
            && cached.hash == hash
        {
            return Ok(cached.data);
        }

        let computed = compute()?;
        if let Ok(data) = serde_json::to_vec(&Cached { data: &computed, hash })
        {
            let _ = fs::create_dir_all(path.parent().expect("invariant"));
            let _ = fs::write(path, data);
        }
        Ok(computed)
    }

    /// Derives a stable entry path from a page identity.
    fn path<K>(&self, namespace: &str, key: &K) -> PathBuf
    where
        K: Hash + ?Sized,
    {
        let mut hasher = DefaultHasher::new();
        BACKLINK_CACHE_VERSION.hash(&mut hasher);
        key.hash(&mut hasher);
        self.directory
            .join(namespace)
            .join(format!("{}.json", hasher.finish()))
    }

    /// Hashes the identity and all inputs that can change an artifact.
    fn hash<I, A>(&self, id: &I, args: &A) -> u64
    where
        I: Hash + ?Sized,
        A: Hash + ?Sized,
    {
        let mut hasher = DefaultHasher::new();
        self.config_hash.hash(&mut hasher);
        id.hash(&mut hasher);
        args.hash(&mut hasher);
        hasher.finish()
    }
}

// ----------------------------------------------------------------------------

impl Mkdocstrings {
    /// Resolves the private settings owned by this pipeline instance.
    pub fn new(config: &Config) -> Self {
        let cache = config.get_cache_dir();
        let backlink_cache = (config.has_markdown_extension(EXTENSION_NAME)
            && config.records_backlinks())
        .then(|| BacklinkCache {
            directory: cache.join("mkdocstrings"),
            config_hash: config.hash,
        });
        Self {
            cache,
            backlink_cache,
            output: config.output_root().clone(),
        }
    }

    /// Installs object inventory generation.
    pub fn setup(&self, dependencies: Dependencies<'_>) {
        let pipeline = self.clone();
        let _ = dependencies.navigation.map(move |_: &Navigation| {
            let cache_path = pipeline.cache.join("objects.inv");
            let cached = fs::read(&cache_path).ok();

            let data = Python::attach(|py| {
                let module = py.import("zensical.compat.mkdocstrings")?;
                module
                    .call_method1("get_inventory", (cached,))?
                    .extract::<Vec<u8>>()
            })?;

            let path = pipeline.output.join(
                &"objects.inv".parse::<SitePath>().expect("static site path"),
            );
            fs::create_dir_all(path.parent().expect("invariant"))?;
            fs::write(path, &data)?;
            fs::create_dir_all(&pipeline.cache)?;
            fs::write(&cache_path, &data)?;
            Ok::<_, anyhow::Error>(())
        });
    }

    /// Creates a visitor when handler backlink rendering is enabled.
    pub fn parser(&self) -> Option<Parser> {
        self.backlink_cache.as_ref().map(|_| Parser::default())
    }

    /// Adds rendered backlinks to the shared HTML edits after collection.
    pub fn apply_backlinks(
        &self, parser: Parser, editor: &mut Editor<'_>, autorefs: &Registry,
        from_url: &str,
    ) -> anyhow::Result<()> {
        let Some(cache) = &self.backlink_cache else {
            return Ok(());
        };
        parser.render(editor, |descriptors| {
            let compute = || {
                let aliases = cache.get_or_compute(
                    "aliases",
                    from_url,
                    descriptors,
                    || {
                        descriptors
                            .iter()
                            .map(|(handler, identifier)| {
                                Self::get_backlink_aliases(handler, identifier)
                            })
                            .collect::<anyhow::Result<Vec<_>>>()
                    },
                )?;
                descriptors
                    .iter()
                    .zip(aliases)
                    .map(|((handler, identifier), aliases)| {
                        Self::render_backlink_placeholder(
                            autorefs, from_url, handler, identifier, aliases,
                        )
                    })
                    .collect::<anyhow::Result<Vec<_>>>()
            };
            if let Some(revision) = autorefs.backlink_revision() {
                cache.get_or_compute(
                    "resolved",
                    from_url,
                    &(revision, descriptors),
                    compute,
                )
            } else {
                compute()
            }
        })
    }

    /// Computes one backlink placeholder from the settled autorefs index.
    fn render_backlink_placeholder(
        autorefs: &Registry, from_url: &str, handler: &str, identifier: &str,
        aliases: Vec<String>,
    ) -> anyhow::Result<String> {
        let mut identifiers = Vec::with_capacity(aliases.len() + 1);
        identifiers.push(identifier.to_string());
        identifiers.extend(aliases);
        identifiers.extend(autorefs.get_aliases(identifier, from_url));
        identifiers.sort();
        identifiers.dedup();
        let backlinks = autorefs.get_backlinks(&identifiers, from_url);
        if backlinks.is_empty() {
            return Ok(String::new());
        }

        Python::attach(|py| {
            py.import("zensical.compat.mkdocstrings")?
                .call_method1("render_backlinks", (handler, backlinks))?
                .extract::<String>()
                .map_err(Into::into)
        })
    }

    /// Gets aliases that only exist in the handler's object model.
    fn get_backlink_aliases(
        handler: &str, identifier: &str,
    ) -> anyhow::Result<Vec<String>> {
        Python::attach(|py| {
            py.import("zensical.compat.mkdocstrings")?
                .call_method1("get_backlink_aliases", (handler, identifier))?
                .extract::<Vec<String>>()
                .map_err(Into::into)
        })
    }
}

// ----------------------------------------------------------------------------
// Tests
// ----------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use std::cell::Cell;

    use tempfile::tempdir;

    use super::BacklinkCache;

    #[test]
    fn backlink_pages_are_reused_without_computing_again() {
        let directory = tempdir().unwrap();
        let cache = BacklinkCache {
            directory: directory.path().to_path_buf(),
            config_hash: 42,
        };
        let descriptors = [("python", "package.Object")];
        let calls = Cell::new(0);

        let first: String = cache
            .get_or_compute("resolved", "reference/", &descriptors, || {
                calls.set(calls.get() + 1);
                Ok("<aside>Guide</aside>".into())
            })
            .unwrap();
        let second: String = cache
            .get_or_compute("resolved", "reference/", &descriptors, || {
                calls.set(calls.get() + 1);
                Ok("rendered twice".into())
            })
            .unwrap();

        assert_eq!(first, "<aside>Guide</aside>");
        assert_eq!(second, first);
        assert_eq!(calls.get(), 1);
    }

    #[test]
    fn backlink_pages_are_scoped_to_configuration() {
        let directory = tempdir().unwrap();
        let descriptors = [("python", "package.Object")];
        let first = BacklinkCache {
            directory: directory.path().to_path_buf(),
            config_hash: 1,
        };
        let second = BacklinkCache {
            directory: directory.path().to_path_buf(),
            config_hash: 2,
        };

        assert_eq!(
            first.path("resolved", "reference/"),
            second.path("resolved", "reference/"),
        );
        let first: String = first
            .get_or_compute("resolved", "reference/", &descriptors, || {
                Ok("first".into())
            })
            .unwrap();
        let second: String = second
            .get_or_compute("resolved", "reference/", &descriptors, || {
                Ok("second".into())
            })
            .unwrap();
        assert_eq!(first, "first");
        assert_eq!(second, "second");
    }
}
